core/src/main/scala/org/bitcoins/core/gcs/GCS.scala

```diff
@@ -0,0 +1,224 @@
+    */
+  def hashToRange(item: ByteVector, f: UInt64, key: ByteVector): UInt64 = {
+    val hash = sipHash(item, key)
+
+    val bigInt = (hash.toBigInt * f.toBigInt) >> 64
+
+    UInt64(bigInt)
+  }
+
+  /**
+    * Hashes the items of a set of items
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#hashing-data-objects]]
+    */
+  def hashedSetConstruct(
+      rawItems: Vector[ByteVector],
+      key: ByteVector,
+      m: UInt64): Vector[UInt64] = {
+    val n = rawItems.length
+    val f = m * n
+
+    val hashedItemsBuilder = Vector.newBuilder[UInt64]
+
+    rawItems.foreach { item =>
+      val setValue = hashToRange(item, f, key)
+      hashedItemsBuilder.+=(setValue)
+    }
+
+    hashedItemsBuilder.result()
+  }
+
+  /**
+    * Converts num to unary (6 = 1111110)
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#golomb-rice-coding]]
+    *
+    * TODO: protect against large inputs which cause OutOfMemoryErrors and even larger ones which fail on toInt
+    */
+  def toUnary(num: UInt64): BitVector = {
+    if (num == UInt64.zero) {
+      BitVector.bits(Vector(false))
+    } else {
+      /*
+       * We use the fact that 2^n - 1 = 111...1 (in binary) where there are n 1 digits
+       */
+      val binUnary = (BigInt(1) << num.toInt) - 1
+      val leftPadded = BitVector(binUnary.toByteArray)
+      val noPadding = dropLeftPadding(leftPadded)
+
+      noPadding.:+(false)
+    }
+  }
+
+  /**
+    * Encodes a hash into a unary prefix and binary suffix
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#golomb-rice-coding]]
+    */
+  def golombEncode(item: UInt64, p: UInt8): BitVector = {
+    val q = item >> p.toInt
+
+    val prefix = toUnary(q)
+
+    val pBits = item.bytes.toBitVector.takeRight(p.toInt)
+
+    prefix ++ pBits
+  }
+
+  /**
+    * Decodes an item off of the front of a BitVector by reversing [[GCS.golombEncode]]
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#golomb-rice-coding]]
+    */
+  def golombDecode(codedItem: BitVector, p: UInt8): UInt64 = {
+    @tailrec
+    def split(vec: BitVector, accum: UInt64): (UInt64, BitVector) = {
+      if (vec.head) {
+        split(vec.tail, accum + UInt64.one)
+      } else {
+        (accum, vec.tail)
+      }
+    }
+
+    val (q, pBits) = split(codedItem, UInt64.zero)
+
+    val sizeWithPadding = (8 - (p.toInt % 8)) + p.toInt
+
+    val pBitsAsBytes = {
+      val withoutRightPaddingOrData = pBits.take(p.toInt)
+      val withLeftPadding = withoutRightPaddingOrData.padLeft(sizeWithPadding)
+      withLeftPadding.toByteVector
+    }
+
+    (q << p.toInt) + UInt64.fromBytes(pBitsAsBytes)
+  }
+
+  @tailrec
+  private def dropLeftPadding(padded: BitVector): BitVector = {
+    if (padded.isEmpty || padded.head) {
+      padded
+    } else {
+      dropLeftPadding(padded.tail)
+    }
+  }
+
+  private def golombDecodeItemFromSet(
```

Comment:
Looks like the `BitVector` returned is surplus bits. Could we add a one line scaladoc?